### ISSUE TITLE
Support S3 compatible storage backend for recorded sessions

### ISSUE THREAD

**Issue by balpert89** [gravitational/teleport#3054]
**What happened**:
I am trying to configure a third party S3-compatible backend storage for recorded sessions which is currently not supported, see reproduce section. Also see https://community.gravitational.com/t/teleport-s3-storage-backend/316 for the community thread.

**What you expected to happen**:
Teleport Auth server starts successfully with S3 third party configured.

**How to reproduce it (as minimally and precisely as possible)**:
Configure the s3-relevant storage parts for a third party s3.
See example config below.

**Environment**:
- Teleport version (use `teleport version`): v4.0.2

**Relevant Debug Logs If Applicable**

Third Party Config:
```
auth_service:
  authentication:
    second_factor: otp
    type: local
  client_idle_timeout: 1h
  cluster_name: t123-sss8d
  disconnect_expired_cert: true
  enabled: true
  listen_addr: 0.0.0.0:3025
  public_addr: teleport-auth-t123:32025
  session_recording: proxy
  tokens:
  - auth,proxy,node:...
proxy_service:
  enabled: false
  kubernetes:
    enabled: false
    listen_addr: ""
    public_addr: null
ssh_service:
  enabled: false
teleport:
  auth_servers:
  - teleport-auth-t123:32025
  auth_token: ...
  connection_limits:
    max_connections: 1000
    max_users: 250
  data_dir: /var/lib/teleport
  log:
    output: stdout
    severity: INFO
  storage:
    # session state
    insecure: false
    peers:
    - mcs-etcd-cluster-client:2379
    prefix: teleport
    tls_ca_file: /etc/tls/private/etcd/etcd-client-ca.crt
    tls_cert_file: /etc/tls/private/etcd/etcd-client.crt
    tls_key_file: /etc/tls/private/etcd/etcd-client.key
    type: etcd
    # recorded sessions
    region: eu-de
    audit_sessions_uri: s3://obs.eu-de.otc.t-systems.com/t123-sss8d-teleport
    access_key: ...
    secret_key: ...
```
Start Up Log
```
Updating certificates in /etc/ssl/certs...
3 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
INFO [PROC]      Generating new host UUID: 5978ba66-b85e-4b49-887f-5c9ae1981b55. service/service.go:547
INFO [MIGRATE]   Importing data from legacy backend. legacy/import.go:53
INFO [MIGRATE]   Successfully imported 0 items. legacy/import.go:66
INFO [PROC:1]    Service diag is creating new listener on 0.0.0.0:3000. service/signals.go:213
INFO [DIAG:1]    Starting diagnostic service on 0.0.0.0:3000. service/service.go:1684
INFO [S3]        Setting up bucket "obs.eu-de.otc.t-systems.com", sessions path "/t123-sss8d-teleport" in region "eu-de". s3sessions/s3handler.go:92
error: "RequestError: send request failed\ncaused by: Head https://s3.eu-de.amazonaws.com/obs.eu-de.otc.t-systems.com: dial tcp: lookup s3.eu-de.amazonaws.com on 172.30.0.10:53: no such host, initialization failed"
```

AWS Configuration
```
auth_service:
  authentication:
    second_factor: otp
    type: local
  client_idle_timeout: 1h
  cluster_name: t123-sss8d
  disconnect_expired_cert: true
  enabled: true
  listen_addr: 0.0.0.0:3025
  public_addr: teleport-auth-t123:32025
  session_recording: proxy
  tokens:
  - auth,proxy,node:...
proxy_service:
  enabled: false
  kubernetes:
    enabled: false
    listen_addr: ""
    public_addr: null
ssh_service:
  enabled: false
teleport:
  auth_servers:
  - teleport-auth-t123.otc:32025
  auth_token: ...
  connection_limits:
    max_connections: 1000
    max_users: 250
  data_dir: /var/lib/teleport
  log:
    output: stdout
    severity: INFO
  storage:
    insecure: false
    peers:
    - mcs-etcd-cluster-client:2379
    prefix: teleport
    tls_ca_file: /etc/tls/private/etcd/etcd-client-ca.crt
    tls_cert_file: /etc/tls/private/etcd/etcd-client.crt
    tls_key_file: /etc/tls/private/etcd/etcd-client.key
    type: etcd
    region: eu-central-1
    audit_sessions_uri: s3://t123-sss8d-teleport
    access_key: ...
    secret_key: ...
```
Start Up Log
```
Updating certificates in /etc/ssl/certs...
3 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
INFO [PROC]      Generating new host UUID: f99741f0-7d8a-4e56-b1ab-084114eb799f. service/service.go:547
INFO [MIGRATE]   Importing data from legacy backend. legacy/import.go:53
INFO [MIGRATE]   Successfully imported 0 items. legacy/import.go:66
INFO [PROC:1]    Service diag is creating new listener on 0.0.0.0:3000. service/signals.go:213
INFO [DIAG:1]    Starting diagnostic service on 0.0.0.0:3000. service/service.go:1684
INFO [S3]        Setting up bucket "t123-sss8d-teleport", sessions path "" in region "eu-central-1". s3sessions/s3handler.go:92
INFO [S3]        Setup bucket "t123-sss8d-teleport" completed. duration:133.303908ms s3sessions/s3handler.go:96
INFO [AUTH]      Updating cluster configuration: StaticTokens([ProvisionToken(Roles=[Auth Proxy Node], Expires=never)]). auth/init.go:245
INFO [AUTH]      Updating cluster configuration: AuthPreference(Type="local",SecondFactor="otp"). auth/init.go:252
INFO [AUTH]      Created namespace: "default". auth/init.go:259
INFO [AUTH]      Auth server is running periodic operations. auth/init.go:399
INFO [CA]        Generating TLS certificate {0x35076a0 0xc0004d4180 CN=f99741f0-7d8a-4e56-b1ab-084114eb799f.t123-sss8d,O=Admin 2029-10-01 14:15:37.157778531 +0000 UTC [teleport-auth-dcc9648f5-4x7x5 teleport-auth-t123.otc.mcs-paas.dev *.teleport.cluster.local teleport.cluster.local]}. common_name:f99741f0-7d8a-4e56-b1ab-084114eb799f.t123-sss8d dns_names:[teleport-auth-dcc9648f5-4x7x5 teleport-auth-t123.otc.mcs-paas.dev *.teleport.cluster.local teleport.cluster.local] locality:[] not_after:2029-10-01 14:15:37.157778531 +0000 UTC org:[Admin] org_unit:[] tlsca/ca.go:176
INFO [PROC]      Admin has obtained credentials to connect to cluster. service/connect.go:377
INFO [PROC:1]    The process has successfully wrote credentials and state of Admin to disk. service/connect.go:417
INFO [PROC:1]    Service auth is creating new listener on 0.0.0.0:3025. service/signals.go:213
INFO [AUTH:1]    Starting Auth service with PROXY protocol support. service/service.go:1039
WARN [AUTH:1]    Configuration setting auth_service/advertise_ip is not set. guessing 10.131.0.73:3025. service/service.go:1114
INFO [AUTH]      Auth service is starting on 0.0.0.0:3025. utils/cli.go:177
[AUTH]    Auth service is starting on 0.0.0.0:3025.
INFO [PROC:1]    Detected that service started and joined the cluster successfully. service/state.go:56
INFO [PROC:1]    The new service has started successfully. Starting syncing rotation status with period 10m0s. service/connect.go:433
```

**Comment by klizhentas** [comment]
@balpert89 what kind of storage is this? I will be working on the new S3 storage integrations in a couple of weeks and can test your storage as well.

**Comment by balpert89** [comment]
Not sure if I understand correctly, the public cloud provider provides a fully aws golang sdk compatible API. The problem is, teleport creates a hostname which forces AWS down my throat: https://s3.eu-de.amazonaws.com/obs.eu-de.otc.t-systems.com/t123-sss8d-teleport

The API endpoint is: obs.eu-de.otc.t-systems.com
The bucket name is: t123-sss8d-teleport

**Comment by balpert89** [comment]
https://github.com/gravitational/teleport/blob/8ce1000ee18c4ce6ae936f24b9d2a73c613d601b/lib/service/service.go#L783

Here is where the Bucket is set, this is set to the URI host, which in the example (`audit_sessions_uri: s3://obs.eu-de.otc.t-systems.com/t123-sss8d-teleport`) would be `obs.eu-de.otc.t-systems.com` - but this is the API endpoint.

**Comment by balpert89** [comment]
@klizhentas @aelkugia I have created a PR (#3057) that tends to the issue, happy for feedback.

**Comment by benarent** [comment]
Fixed via #3054
